### PR TITLE
Added Span overloads for Socket.SendFile

### LINKED ISSUE #43846: Add Span overloads for Socket.SendFile
## Proposed API

```csharp
namespace System.Net.Sockets
{
    public class Socket
    {
        // Existing API
        // public void SendFile(string fileName, byte[] preBuffer, byte[] postBuffer, System.Net.Sockets.TransmitFileOptions flags);

        public void SendFile(string fileName, ReadOnlySpan<byte> preBuffer, ReadOnlySpan<byte> postBuffer, System.Net.Sockets.TransmitFileOptions flags);
    }
}
```

### PR DESCRIPTION
**Pull request by gfoidl**
Fixes https://github.com/dotnet/runtime/issues/43846

---

This PR is split-out of https://github.com/dotnet/runtime/pull/45132 -- see https://github.com/dotnet/runtime/pull/45132#issuecomment-763590998 for context.

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1277]: Does this handle null values for pre/postBuffer correctly?

Do we have tests for that?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1288]: This comment is wrong in multiple ways -- it's not a byte array, it can't be null.

Is this just copied from existing docs? If so, please add it to the existing overloads, and then revise this one to be appropriate for the new overload.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1297]: BTW do we have a test for this? I.e. that trying to do SendFile over UDP throws NotSupportedException? If not we should.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1298]: Similarly, do we have test that we get InvalidOperationException if you try to do this with a non blocking socket?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: I'm not sure what happens when you pin an empty Span. Do you get null? Do you get a valid pointer to some random empty array that actually gets pinned? If the latter, that seems bad

Seems like we should ensure that we pass null to TransmitFile for empty spans.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1277]: A `null` array will become an empty span, which is checked in the control flow.

No explicit test, but it's covered by https://github.com/dotnet/runtime/blob/5a4483eb1c0918542298d1864b7233f962952880/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L137-L147

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1288]: > Is this just copied from existing docs?

Ui, yes and now for this PR from the old. I feel ashamed that this happened. Thanks.

**Comment by geoffkizer** [comment]
Tests?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: Just chatted with @stephentoub -- doing (fixed byte* p = someSpan) will indeed give you null if the span is empty.

Would probably be good to assert this too, though.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1297]: There's a test for https://github.com/dotnet/runtime/blob/5a4483eb1c0918542298d1864b7233f962952880/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L98

I'll add a test for UDP.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1288]: We're in kind of a weird space currently for changes like this, i.e. adding a new overload to a method that has docs on MSDN but not in the source. Best policy is to copy the existing docs for the existing overloads into the source manually, then add/adapt that for the new overload(s) too.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: >  pin an empty Span.

Results in a null-pointer / IntPtr.Zero for this usage after the cast. This is safe.

**Comment by gfoidl** [comment]
> Tests?

Tests for the array-based implementation are there, and as these forward to the span-based implementation that's covered.
Explicit span tests are a duplication, but I can add those if you want.

**Comment by geoffkizer** [comment]
We should add explicit span tests too. Yeah it's duplicative, but that's fine; better to be duplicative than to miss something.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: Didn't see the new comment before writing mine.

The casts https://github.com/dotnet/runtime/blob/878fdf3a3b8f9f8edc65c1177ea4592424c54dc0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L299 make it `IntPtr.Zero` and for the length `0` will be passed to `TransmitFileHelper` where these lengths are checked https://github.com/dotnet/runtime/blob/878fdf3a3b8f9f8edc65c1177ea4592424c54dc0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L1036 so this operation is safe.

I don't understand what should be asserted here.

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: Yeah, you are right, we are checking the length again in TransmitFileHelper so it doesn't matter either way.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:294]: > doing (fixed byte* p = someSpan) will indeed give you null if the span is empty.

Yup.  Just sharing in the name of completeness, that logic lives here:
https://github.com/dotnet/runtime/blob/8c76312b621e2cdb22ca0ee603c5b0e1919134a2/src/libraries/System.Private.CoreLib/src/System/Span.cs#L255-L262
When you write:
```C#
fixed (byte* p = someSpan)
```
the C# compiler treats that as equivalent to if you wrote:
```C#
fixed (byte* p = &someSpan.GetPinnableReference())
```
thus invoking the previously cited code, which explicitly checks for 0-length and returns null in that case.

Note that this is not the case if you instead used:
```C#
fixed (byte* p = &MemoryMarshal.GetReference(someSpan))
```
whose sole reason to exist is to not perform that 0-length check.

**Comment by antonfirsov** [comment]
To support all the refactoring described in https://github.com/dotnet/runtime/pull/45132#issuecomment-763880519 and #43845, we should refactor `SendFileTest` to derive from `SocketTestHelperBase<T>` similarly to #46862.

I strongly suggest to do it in this PR.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1298]: There's https://github.com/dotnet/runtime/blob/5a4483eb1c0918542298d1864b7233f962952880/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L213 

Trying something like
```c#
using var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
listener.BindToAnonymousPort(IPAddress.Loopback);
listener.Listen(1);

client.Connect(listener.LocalEndPoint);
using Socket server = listener.Accept();
server.ForceNonBlocking(force: true);

Assert.Throws<InvalidOperationException>(() => server.SendFile(null));
```
won't throw the exception (i.e. that test fails) on Windows and Linux.
After `ForceNonBlocking` both `_willBlock` and `_willBlockInternal` are `false`, so https://github.com/dotnet/runtime/blob/16e6e9530687b34cc93a85d4043093f06f9b7eff/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4045-L4051 won't throw.

**Comment by antonfirsov** [comment]
@gfoidl `FileCleanupTestBase.GetTestFilePath()` is used to create and auto-cleanup temporary files. The `TempFile` utility is an equivalent tool for this, and it doesn't need subclassing:
https://github.com/dotnet/runtime/blob/master/src/libraries/Common/tests/System/IO/TempFile.cs

`SocketTestHelperBase<T>` is a utility to avoid over-duplicating the same test logic across all the Sync / APM (Begin, End) / SocketAsyncEventArgs / Task test variants. This makes it easier to ensure that the cases which are valid for all those variants are indeed tested for all of them.

What I wanted was to avoid bloating the current `SendFile.cs` further by coding the same tests for the new `Span<byte>` overloads (just to delete them a few weeks later). Nevertheless, if a separate PR is easier for you, I'm fine with that, and I may also take that work after finishing #47229.

**Comment by gfoidl** [comment]
Thanks for the explaination!
Please let's sheperd this PR to merge, then do the test stuff where I'd really appreciate

> I'm fine with that, and I may also take that work after finishing #47229.

to see how it's done (and use that knowledge then for the SendFileAsync tests).

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: I don't mean ForceNonBlocking. That puts the socket into sync-over-async mode.

I mean literally a nonblocking socket, i.e. socket.Blocking = false.

**Comment by geoffkizer** [comment]
I'm fine to defer the `SocketTestHelperBase<T>` work until after we merge this PR, but we should do it before we add the Task-based SendFile overloads.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Tried this too, same result where the above `if` in ValidateBlockingMode won't throw.

grafik

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Huh. It seems like we don't actually prevent you from doing a sync SendFile in non-blocking mode.

Does it actually work? What happens if you send a really big file, does it block or return E_WOULDBLOCK or what?

I think we probably should be preventing this, but it looks like we are not currently...

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: The following program sends 1000 MiB and runs just fine.
With the debugger it can be verifyed that it blocks until the send is finished (will step over once the client has read the sent data). I.e. it's the same behavior as when `server.Blocking = true`.

<details>
  <summary>SendFile non-blocking 1000 MiB</summary>

```c#
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;

using var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
listener.BindToAnonymousPort(IPAddress.Loopback);
listener.Listen(1);

client.Connect(listener.LocalEndPoint);
using Socket server = listener.Accept();
server.Blocking = false;

Task clientTask = RunClientAsync();

string fileName = Path.GetTempFileName();
try
{
    const long FileLength = 100 * 1024 * 1024;
    using (FileStream fs = File.Create(fileName))
    {
        fs.SetLength(FileLength);
    }

    Debug.Assert(!server.Blocking);

    server.SendFile(fileName);
    server.Close();
}
catch (SocketException ex)
{
    Console.WriteLine(ex.Message);
    Console.WriteLine(ex.SocketErrorCode);
}
finally
{
    File.Delete(fileName);
}

await clientTask;

Task RunClientAsync() => Task.Run(() =>
{
    byte[] buffer = new byte[4096];
    long totalReceived = 0;
    int received = 0;
    while ((received = client.Receive(buffer)) > 0)
    {
        Console.Write('.');
        totalReceived += received;
    }

    Console.WriteLine($"\nClient received {totalReceived / 1024 / 1024} MiB");
});

internal static class SocketTestExtensions
{
    // Binds to an IP address and OS-assigned port. Returns the chosen port.
    public static int BindToAnonymousPort(this Socket socket, IPAddress address)
    {
        socket.Bind(new IPEndPoint(address, 0));
        return ((IPEndPoint)socket.LocalEndPoint).Port;
    }
}
```
</details>

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Is that on Linux or Windows?

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Windows. Will try Linux soon

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: On linux it throws:
```
Message: Resource temporarily unavailable
SocketErrorCode: WouldBlock
```

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: There is something strange. When in the above code 
```diff
server.Blocking = false;
+DumpState(server);
```
```c#
static void DumpState(Socket socket)
{
    BindingFlags flags = BindingFlags.Instance | BindingFlags.NonPublic;
    FieldInfo _willBlock = typeof(Socket).GetField("_willBlock", flags);
    FieldInfo _willBlockInternal = typeof(Socket).GetField("_willBlockInternal", flags);

    object willBlock = _willBlock.GetValue(socket);
    object willBlockInternal = _willBlockInternal.GetValue(socket);

    Console.WriteLine($"Blocking: {socket.Blocking}\nwillBlock: {willBlock}\nwillBlockInternal: {willBlockInternal}\n");
}
```
Then on Windows it's the same behaviour. It runs fine whether `server.Blocking` is true/false.

On Linux it gets flaky with this line. Means on some runs it fails immediately (expected), on some runs it finished without failure or it fails somewhere in the middle.
Seems there's a race-condition anywhere?

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Interesting.

We shouldn't block this PR on this issue, since it already exists. I'll file a separate issue.

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: That's very strange. I'm not surprised it's flaky on Linux, but I am surprised that whether you call "DumpState" or not determines whether it is flaky. Very odd.

**Comment by geoffkizer** [comment]
ok -- are we ready to merge then? @antonfirsov any remaining concerns?

**Comment by gfoidl** [comment]
I have no more changes planned for this PR. So ready to merge from my side.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1319]: Issue here: https://github.com/dotnet/runtime/issues/47287

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs:112]: `[PlatformSpecific]` is enough to make this conditional, `[ConditionalTheory]` is used when you need to evaluate `bool` members at runtime.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs:132]: Wouldn't it make sense to also check `SocketException.SocketErrorCode`?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1294]: Looks like we don't have test for the `FileNotFoundException` case.

**Review by antonfirsov — APPROVED** [review]
LGTM. I have some comments regarding tests, but let's ignore them, since it doesn't make sense to do alterations to test code we are about to completely refactor anyways.

I want to merge as soon as the outerloop run finishes.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1294]: Indeed, that's missing.

As the tests will be refactored, should we file an issue for this and where these points get collected? Or won't we forget about this anyway?

Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1294]: If one of us manages to PR the test refactor within a week, I wouldn't bother with an issue, otherwise it would be wise to start tracking it.

My main fear is that more extensive testing will discover a bunch of corner cases and platform inconsistencies (or even actual bugs) we'll have to deal with. This is what made things hard with my other PR.

**Comment by antonfirsov** [comment]
OuterLoop test failures are unrelated:
#41606
#1724
https://github.com/dotnet/runtime/issues/21631#issuecomment-766691502
#47103
#46439 (test run before merging the fix)

**Comment by antonfirsov** [comment]
@gfoidl FYI I have started refactoring the SendFile tests, if everything goes well, expect a PR within a day.
https://github.com/dotnet/runtime/compare/master...antonfirsov:sockets/SendFile-test-refactor?expand=1

**Comment by gfoidl** [comment]
Thanks for the update -- happy to have a look at the PR (then start the async versions of sendfile).

**Pull request merged**
